File: leetcodejava/src/main/java/com/leetcode/SixOne.java
package com.leetcode;

import com.entity.ListNode;

/**
 * @author: John
 * @date: 2020-02-05 17:18
 * @description: 61
 **/
public class SixOne {

//    public ListNode rotateRight(ListNode head, int n) {
//        if (head == null || head.next == null) {
//            return head;
//        }
//        ListNode dummy = new ListNode(0);
//        dummy.next = head;
//        ListNode fast = dummy, slow = dummy;
//        int i;
//        for (i = 0; fast.next != null; i++) {
//            fast = fast.next;
//        }
//        for (int j = i-n%i; j>0; j--) {
//            slow = slow.next;
//        }
//        // there can generate infinite loop
//        fast.next = dummy.next;
////        System.out.println(dummy.next.val);
//        dummy.next = slow.next;
//        slow.next = null;
//        return dummy.next;
//    }

//    public ListNode rotateRight(ListNode head, int k) {
//        if (head == null) {
//            return null;
//        }
//        if (head.next == null) {
//            return head;
//        }
//        ListNode old_tail = head;
//        int n;
//        for (n=1; old_tail.next != null; n++) {
//           old_tail = old_tail.next;
//        }
//        old_tail.next = head;
//        ListNode new_tail = head;
//        for (int i = 0; i < n-k%n-1; i++) {
//            new_tail = new_tail.next;
//        }
//        ListNode new_head = new_tail.next;
//        new_tail.next = null;
//        return new_head;
//    }

//    public ListNode rotateRight(ListNode head, int n) {
//        if (head == null || head.next == null) {
//            return head;
//        }
//        ListNode dummy = new ListNode(0);
//        dummy.next = head;
//        ListNode fast = dummy, slow = dummy;
//        int i;
//        for (i = 0; fast.next != null; i++) { // Get the total length
//            fast = fast.next;
//        }
//
//        for (int j = i - n % i; j > 0; j--) {   // get the i - n % i th node
//            slow = slow.next;
//        }
//        fast.next = dummy.next; // Do the rotation
//        dummy.next = slow.next;
//        slow.next = null;
//        return dummy.next;
//    }

    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) {
            return head;
        }
        int len = 1; // number of nodes
        ListNode newNode = head, tail = head;

        while (tail.next != null) { // get the number of nodes in the list
            tail = tail.next;
            len++;
        }
        tail.next = head; // circle the link

        if ((k %= len) > 0) {
           for (int i = 0; i < len - k; i++) {
              tail = tail.next; //  the tail node is the (len - k) th node(1st node is head)
           }
        }
        newNode = tail.next;
        tail.next = null;
        return newNode;
    }

    public static void main(String[] args) {
        ListNode node1 = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        ListNode node5 = new ListNode(5);

        node4.next = node5;
        node3.next = node4;
        node2.next = node3;
        node1.next = node2;

        SixOne sixOne = new SixOne();
        sixOne.rotateRight(node1, 2);
    }
}

File: leetcodecpp/leetcode/SixOne.cpp
//
// Created by John on 2020/2/5.
//

#include "../algorithm/LeetCodeInterview.h"
#include "stack"

class Solution{
//public:
//    static ListNode* rotateRight(ListNode * head, int k) {
//        if (!head) {
//            return head;
//        }
//        int len = 1;
//        ListNode *newH, *tail;
//        newH = tail = head;
//        while (tail->next) {
//            tail = tail->next;
//            len++;
//        }
//        tail->next = head;
//        if (k %= len) {
//            for (auto i = 0; i < len -k; i ++) {
//                tail = tail->next;
//            }
//        }
//        newH = tail->next;
//        tail->next = NULL;
//        return newH;
//    }

//public:
//    static ListNode* rotateRight(ListNode * head, int k) {
//        if (!k || !head || !head->next) {
//            return head;
//        }
//        ListNode *fast = head;
//        int lens = 1;
//        while (k--) {
//            if (fast->next) {
//                fast = fast->next;
//                lens++;
//            } else {
//                fast = head;
//                k %= lens;
//            }
//        }
//        if (fast == head) {
//            return head;
//        }
//        ListNode *slow = head;
//        while(fast->next) {
//            fast = fast->next;
//            slow = slow->next;
//        }
//        ListNode *newhead = slow->next; // next is fast
//        slow->next = NULL;
//        fast->next = head;
//        return newhead;
//    }
//public:
//    ListNode* rotateRight(ListNode * head, int k) {
//        if (!head) {
//            return head;
//        }
//        int len = 1; // number of nodes
//        ListNode *newH, *tail;
//        newH = tail = head;
//
//        while (tail->next) { // get the number of nodes in the list
//            tail = tail->next;
//            len++;
//        }
//        tail->next = head; // circle the link
//
//        if (k %= len) {
//            for (auto i = 0; i < len - k; i++) {
//                tail->next;     // the tail node is the(len - k) -th node(1st node is head)
//            }
//        }
//        newH = tail->next;
//        tail->next = NULL;
//        return newH;
//    }

public:
    static ListNode* rotateRight(ListNode *head, int k) {
        if (!head) {
            return head;
        }
        int len = 1;
        ListNode *p = head;
        while (p->next) {
            len++;
            p = p->next;
        }
        p->next = head;

        if (k %= len) {
            for (int i = 0; i < len - k; ++i, p = p->next);
        }
        ListNode* newHead = p->next;
        p->next = NULL;
        return newHead;
    }

    // myself
//    static ListNode* rotateRight(ListNode *head, int k) {
//        if (!head) {
//            return head;
//        }
//        ListNode *dummy = new ListNode(0);
//        dummy->next = head;
//        int total = 0;
//
//        while (dummy->next) {
//            total++;
//            dummy = dummy->next;
//        }
//        dummy->next = head;
//
//        if (k / total) {
//            k = (total - (k % total) + 1);
//        } else {
//            k = (total - k + 1);
//        }
//
//        ListNode *tail;
//        while (--k >= 0) {
//            if (k == 0) {
//                tail = dummy;
//            }
//            dummy = dummy->next;
//        }
//        tail->next = NULL;
//        return dummy;
//    }
};


/**
 * Input: head = [1,2,3,4,5], k = 2
 * [0,1,2], k = 4
 * [2,0,1]
 * 2 0 1
 * 1 2 0
 * 0 1 2
 * 2 0 1
 */

//int main() {
////    cout << (2 % 5) << endl;
//
////    ListNode * node1 = new ListNode(1);
////    ListNode * node2 = new ListNode(2);
////    ListNode * node3 = new ListNode(3);
////    ListNode * node4 = new ListNode(4);
////    ListNode * node5 = new ListNode(5);
////
////    node4->next = node5;
////    node3->next = node4;
////    node2->next = node3;
////    node1->next = node2;
////    Solution::rotateRight(node1, 2);
//
////        cout << 4 % 3 << endl;
//
//
//    ListNode * zero1 = new ListNode(0);
//    ListNode * node1 = new ListNode(1);
//    ListNode * node2 = new ListNode(2);
//
//    node1->next = node2;
//    zero1->next = node1;
//    Solution::rotateRight(zero1, 4);
//}